File: src/main/java/de/mrapp/textmining/util/ngram/NGram.java
/*
 * Copyright 2017 Michael Rapp
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package de.mrapp.textmining.util.ngram;

import org.jetbrains.annotations.NotNull;

import java.io.Serializable;

import static de.mrapp.util.Condition.*;

/**
 * A n-Gram, which consists of a sequence of characters (token), taken from a longer text or word.
 *
 * @author Michael Rapp
 * @since 1.0.0
 */
public class NGram implements Serializable, Cloneable {

    /**
     * The constant serial version UID.
     */
    private static final long serialVersionUID = -5907278359683181663L;

    /**
     * The n-gram's token.
     */
    private final String token;

    /**
     * The position of the n-gram's token in the original text or word.
     */
    private final int position;

    /**
     * Creates a new n-gram, which consists of a sequence of characters (token).
     *
     * @param token    The token of the n-gram as a {@link String}. The token may neither be null,
     *                 nor empty
     * @param position The position of the n-gram's token in the original text or word as an {@link
     *                 Integer} value. The position must be at least 0
     */
    public NGram(@NotNull final String token, final int position) {
        ensureNotNull(token, "The token may not be null");
        ensureNotEmpty(token, "The token may not be null");
        ensureAtLeast(position, 0, "The position must be at least 0");
        this.token = token;
        this.position = position;
    }

    /**
     * Returns the token of the n-gram.
     *
     * @return The token of the n-gram as a {@link String}. The token may neither be null, nor empty
     */
    @NotNull
    public final String getToken() {
        return token;
    }

    /**
     * Returns the position of the n-gram's token in the original text or word.
     *
     * @return The position of the n-gram's token as an {@link Integer} value. The position must be
     * at least 0
     */
    public final int getPosition() {
        return position;
    }

    /**
     * Returns the length of the n-gram's token.
     *
     * @return The length of the n-gram's token as an {@link Integer} value
     */
    public final int length() {
        return token.length();
    }

    @Override
    public final NGram clone() {
        return new NGram(token, position);
    }

    @Override
    public final String toString() {
        return "NGram [token=" + token + ", position=" + position + "]";
    }

    @Override
    public final int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + token.hashCode();
        result = prime * result + position;
        return result;
    }

    @Override
    public final boolean equals(final Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        NGram other = (NGram) obj;
        return token.equals(other.token) && position == other.position;
    }

}

File: src/main/java/de/mrapp/textmining/util/tokenizer/NGram.kt
/*
 * Copyright 2017 - 2019 Michael Rapp
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package de.mrapp.textmining.util.tokenizer

import de.mrapp.textmining.util.Token
import de.mrapp.util.Condition.ensureAtLeast
import de.mrapp.util.Condition.ensureEqual
import de.mrapp.util.Condition.ensureNotEmpty
import java.util.*

/**
 * A n-Gram, which consists of a sequence of characters, taken from a longer text or word.
 *
 * @property n         The degree of the n-gram
 * @property token     The token of the substring
 * @property positions A set that contains the position(s) of the substring's token in the original
 *                     text
 * @author Michael Rapp
 * @since 1.2.0
 */
data class NGram(val n: Int, override var token: CharSequence,
                 override val positions: SortedSet<Int> = TreeSet()) : Token {

    init {
        ensureAtLeast(n, 1, "The degree must be at least 1")
        ensureNotEmpty(token, "The token may not be empty")
        ensureEqual(positions.count { it < 0 }, 0, "All positions must be at least 0")
    }

    /**
     * Creates a new substring, which consists of a sequence of characters, taken from a longer
     * text.
     *
     * @param n The degree of the n-Gram
     * @param token The token of the substring
     * @param positions An array, which contains the position(s) of the substring's token in the
     * original text
     */
    constructor(n: Int, token: CharSequence, vararg positions: Int) :
            this(n, token, positions.toList())

    /**
     * Creates a new substring, which consists of a sequence of characters, taken from a longer
     * text.
     *
     * @param n The degree of the n-Gram
     * @param token The token of the substring
     * @param positions A collection, which contains the position(s) of the substring's token in the
     * original text
     */
    constructor(n: Int, token: CharSequence, positions: Collection<Int>) :
            this(n, token, TreeSet(positions))

    override fun addPosition(position: Int) {
        ensureAtLeast(position, 0, "The position must be at least 0")
        this.positions.add(position)
    }

    override fun toString() = token.toString()

}
